# EXERCISE WALKTHROUGH

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
data_path = r'../2_Data/'
df_orders = pd.read_csv(os.path.join(data_path, '1_Original_Data', 'orders.csv'), index_col = None)
df_prods = pd.read_csv(os.path.join(data_path, '1_Original_Data', 'products.csv'), index_col = None)

In [3]:
df_orders.drop(columns = ['eval_set'], inplace=True)

In [4]:
df_orders['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [5]:
df_orders.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [6]:
df_orders.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [7]:
df_orders['order_id'] = df_orders['order_id'].astype('str')
df_orders.describe()

user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01        2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [8]:
df_orders['order_id'].dtype

dtype('O')

In [9]:
df_deps = pd.read_csv(os.path.join(data_path, '1_Original_Data', 'departments.csv'), index_col = False)
df_deps.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [10]:
df_deps.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [11]:
df_deps_t = df_deps.T
df_deps_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [12]:
df_deps_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [13]:
new_header = df_deps_t.iloc[0]

In [14]:
new_header

0    department
Name: department_id, dtype: object

In [15]:
df_deps_t_new = df_deps_t[1:]
df_deps_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [16]:
df_deps_t_new.columns = new_header
df_deps_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [17]:
data_dict = df_deps_t_new.to_dict('index')
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [18]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [19]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [20]:
df_snacks =  df_prods[df_prods['department_id']==19]
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# TASK

### Step 2 - change datatype to another column not needed in statistical analysis


In [21]:
df_orders['user_id'] = df_orders['user_id'].astype('str')
df_orders.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

### Step 3 - change a column name to a more meaningful name

In [22]:
df_orders.rename(columns = {'order_number' : 'customer_sequential_order_number'}, inplace = True)
df_orders.head()

order_id user_id  customer_sequential_order_number  orders_day_of_week  \
0  2539329       1                                 1                   2   
1  2398795       1                                 2                   3   
2   473747       1                                 3                   3   
3  2254736       1                                 4                   4   
4   431534       1                                 5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

### Step 4 - What's the busiest hour of the day to place orders?

In [23]:
orders_per_day = df_orders['order_hour_of_day'].value_counts()
print(orders_per_day)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64


The answer is 10 a.m.

### Step 5 - Check which department has the id of 4 - I am making use of the dictionary already created

In [24]:
print(data_dict.get('4'))

{'department': 'produce'}


### Step 6 - create a subset of breakfast products, looking above we can see it has the id of 14.

In [25]:
df_bkfast = df_prods[df_prods['department_id']==14]
df_bkfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

### Step 7 - create a subset that includes all products used for dinner parties, comprised by:
* alcohol       (5)
* deli         (20)
* beverages     (7)
* meat seafood (12)

In [26]:
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

### Step 8 - check how many rows the previously created df has

In [27]:
df_dinner_parties.shape

(7650, 5)

The answer is 7650 rows

### step 9 - extract info from user with user_id == 1

In [28]:
#since we changed datatype, user_id is longer int64, but string, hence '1' instead of 1
df_user_id_1 = df_orders.loc[df_orders['user_id']=='1']
df_user_id_1

order_id user_id  customer_sequential_order_number  orders_day_of_week  \
0   2539329       1                                 1                   2   
1   2398795       1                                 2                   3   
2    473747       1                                 3                   3   
3   2254736       1                                 4                   4   
4    431534       1                                 5                   4   
5   3367565       1                                 6                   2   
6    550135       1                                 7                   1   
7   3108588       1                                 8                   1   
8   2295261       1                                 9                   1   
9   2550362       1                                10                   4   
10  1187899       1                                11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

### step 10 - summary statistics about this user_id

In [29]:
df_user_id_1.describe()

customer_sequential_order_number  orders_day_of_week  \
count                         11.000000           11.000000   
mean                           6.000000            2.636364   
std                            3.316625            1.286291   
min                            1.000000            1.000000   
25%                            3.500000            1.500000   
50%                            6.000000            3.000000   
75%                            8.500000            4.000000   
max                           11.000000            4.000000   

       order_hour_of_day  days_since_prior_order  
count          11.000000               10.000000  
mean           10.090909               19.000000  
std             3.477198                9.030811  
min             7.000000                0.000000  
25%             7.500000               14.250000  
50%             8.000000               19.500000  
75%            13.000000               26.250000  
max            16.000000               30.000000

We can see that this user:
* Never shops on Thursday, Friday or Saturday
* places his orders between 7 a.m. and 4 p.m.
* one time, he shopped twice in the same day, but normally takes between 14 and 30 days to shop again

### step 11 - export orders wrangled to a csv file on data prepared folder

In [30]:
df_orders.to_csv(os.path.join(data_path,'2_Prepared_Data', 'orders_wrangled.csv'), index=False)

### step 12 - export reworked departments dataframe to a csv file on data prepared folder

In [31]:
df_deps_t_new.to_csv(os.path.join(data_path,'2_Prepared_Data', 'departments_wrangled.csv'))